### ISSUE TITLE
[API Proposal]: A more nuanced story for Dependency Injection

### ISSUE THREAD

**Issue by vukovinski** [dotnet/runtime#83666]
### Background and motivation

When building, designing and maintaining large applications with multiple team members or when doing the same with a large number of microservices projects that share a common library, a more nuanced approach to dependency injection is needed.

An ability to configure - not a scope, but a context of services provided to a logical application layer or module would come in handy.

Currently the Microsoft.Extensions.DependencyInjection and its *.Abstractions library provide a robust yet simple notion of dependency injection: configure once, then use everywhere. The problem is especially compounded as one of the lifecycle concepts supported has a special relationship with ASP.NET Core, namely the **scoped** lifetime, which is not a problem in of itself, as the DI library was largely refactored out of ASP.NET and other products due to the nature of history.

On a side note, a large number of libraries hook into the DI library by providing extensions methods to register their services for provision throughout the application - an approach somewhat followed also in this proposal.

We propose a notion of an IServiceProvider, that could be registered multiple times into the root - or app level IServiceProvider, and be configured for and requested by different contexts. We use a generic type parameter as they key for a context. Note that we do not support variance in the type paramater.

### API Proposal

```csharp
namespace Microsoft.Extensions.DependencyInjection.Contexts;

// the gist
public interface IContextualServiceProvider<Context>
{
    IServiceProvider ServiceProvider { get; }
}

public class ContextualServiceProvider<Context> : IContextualServiceProvider<Context>
{
    public IServiceProvider ServiceProvider { get; }

    public ContextualServiceProvider(IServiceProvider serviceProvider) => this.ServiceProvider = serviceProvider;
}

// for extension purposes only
public interface IContextualServiceProviderFactory<Context>
{
   IContextualServiceProvider<Context> CreateNew();
}
```

### API Usage

```csharp
// ServiceContexts.cs
public class ApiControllerContext {}
public class InfrastructureContext {}

// ApiController.cs
public class ApiController : Controller
{
    private readonly IServiceProvider AppServiceProvider;
    private readonly IContextualServiceProvider<ApiControllerContext> ApiControllerServiceProvider;
    private readonly IContextualServiceProvider<InfrastructureContext> InfrastructureServiceProvider;
}

// Program.cs
builder.Services.ConfigureServices(services =>
    services.AddContext<ApiControllerContext>(ConfigureApiControllerContext);
    services.AddContext<InfrastructureContext>(ConfigureInfrastructureContext);
);
```

### Alternative Designs

- a string key could be used instead of a generic type parameter
- a more modular approach using System.ComponentModel could be emphasised
- modularization could be achieved by authoring a composition root for a context

### Risks

- may introduce confusion regarding AddScoped vs AddContext
- may be to opinionated and best be left to users to implement their own solution

**Comment by steveharter** [comment]
>Alternative Designs
>> a string key could be used instead of a generic type parameter

FYI this is covered by https://github.com/dotnet/runtime/issues/64427 which is being evaluated for .NET 8.

**Comment by steveharter** [comment]
cc @benjaminpetit

**Comment by vukovinski** [comment]
Closed with #64427
